othello/apps/games/tasks.py

```diff
@@ -0,0 +1,200 @@
+    print("sending" + event_type)
+    task_logger.debug(f"sending {event_type}")
+    async_to_sync(get_channel_layer().group_send)(
+        game.channels_group_name, {"type": event_type, "object_id": object_id}
+    )
+
+
+def ping(game):
+    if game.is_tournament:
+        return True
+    game.refresh_from_db()
+    return (timezone.now() - game.last_heartbeat).seconds < settings.CLIENT_HEARTBEAT_INTERVAL * 2
+
+
+def delete_game(game):
+    if not game.is_tournament:
+        game.delete()
+
+
+@shared_task
+def run_game(game_id):
+    try:
+        game = Game.objects.get(id=game_id)
+        yourself = Submission.objects.get(user__username="Yourself")
+    except Game.DoesNotExist:
+        logger.error(f"Trying to play nonexistent game ({game_id}")
+        return "game not found"
+
+    mod, time_limit = Moderator(), game.time_limit
+    game_over = False
+    file_deleted = None
+
+    try:
+        black_runner = (
+            YourselfRunner(game, settings.YOURSELF_TIMEOUT)
+            if game.black == yourself
+            else PlayerRunner(game.black, settings.JAILEDRUNNER_DRIVER)
+        )
+    except OSError:
+        logger.error(f"Cannot find submission code file! {game.black.code.path}")
+        black_runner = None
+        file_deleted = game.errors.create(
+            player=Player.BLACK.value,
+            error_code=ServerError.FILE_DELETED.value[0],
+            error_msg=ServerError.FILE_DELETED.value[1],
+        )
+    try:
+        white_runner = (
+            YourselfRunner(game, settings.YOURSELF_TIMEOUT)
+            if game.white == yourself
+            else PlayerRunner(game.white, settings.JAILEDRUNNER_DRIVER)
+        )
+    except OSError:
+        logger.error(f"Cannot find submission code file! {game.white.code.path}")
+        white_runner = None
+        file_deleted = game.errors.create(
+            player=Player.WHITE.value,
+            error_code=ServerError.FILE_DELETED.value[0],
+            error_msg=ServerError.FILE_DELETED.value[1],
+        )
+
+    if file_deleted is not None:
+        game.forfeit = False
+        game.outcome = "T"
+        game.playing = False
+        game.save(update_fields=["forfeit", "outcome", "playing"])
+        send_through_game_channel(game, "game.error", file_deleted.id)
+        raise RuntimeError("Cannot find a submission code file!")
+
+    error = 0
+    with black_runner as player_black, white_runner as player_white:
+        last_move = game.moves.create(board=INITIAL_BOARD, player="-", possible=[26, 19, 44, 37])
+        send_through_game_channel(game, "game.update", game_id)
+        exception = None
+
+        while not mod.is_game_over():
+            if not ping(game):
+                game.playing = False
+                game.outcome = "T"
+                game.forfeit = False
+                game.save(update_fields=["playing", "outcome", "forfeit"])
+                delete_game(game)
+                return "no ping"
```

Comment:
I'd rather log the error and then just return `None`.

Reply:
Yeah, the reason I changed it from returning `None` to returning a string is b/c the return value will show up in Celery Task Results in the django admin interface. It would be nice to go to the admin dashboard and see what the task returned so I could have something to start off of in case I need to investigate an error/debug

Reply:
This would be much better done by implementing a sane logging system and just logging `DEBUG` messages.